packaging/metadata.py

```diff
@@ -0,0 +1,386 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+import dataclasses
+import math
+import sys
+from email import message_from_bytes
+from email.headerregistry import Address, AddressHeader
+from email.message import EmailMessage
+from email.policy import EmailPolicy, Policy
+from functools import reduce
+from inspect import cleandoc
+from itertools import chain
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Collection,
+    Dict,
+    Iterable,
+    Iterator,
+    Mapping,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from .requirements import InvalidRequirement, Requirement
+from .specifiers import SpecifierSet
+from .version import Version
+
+T = TypeVar("T", bound="CoreMetadata")
+A = TypeVar("A")
+B = TypeVar("B")
+
+if sys.version_info[:2] >= (3, 8) and TYPE_CHECKING:  # pragma: no cover
+    from typing import Literal
```

Comment:
Rather than have things type check differently on different Python versions, you can just use:
```
if TYPE_CHECKING:
    from typing_extensions import Literal
```
You don't need to actually install typing_extensions for this to work, since a) it's in the TYPE_CHECKING block, b) type checkers treat typing_extensions as part of the stdlib, so they always know about its existence.

Reply:
Hi @hauntsaninja, thank you for the tip! That sounds very nice... However when I try to run `mypy` with Python 3.7, I get a series of errors, including:

```bash
packaging/metadata.py:41: error: Module "typing" has no attribute "Literal"; maybe "_Literal"?
```

Are you sure about type checkers treating `typing_extensions` as part of the stdlib?

Reply:
Yes, I'm quite sure (I help maintain typeshed and mypy). That error looks like you might not have changed the import to use `typing_extensions` instead of `typing`?

The following works for me (where `emptyenv` does not have `typing_extensions` installed in it):
```
python3.7 -m mypy -c 'from typing_extensions import Literal' --python-version 3.7 --python-executable emptyenv/bin/python
```

To be explicit, the options are:
1)
```
# does not require runtime dependency on typing_extensions
# works because all type checkers special case typing_extensions
if TYPE_CHECKING:
    from typing_extensions import Literal
    NormalizedDynamicFields = Literal[...]
else:
    NormalizedDynamicFields = str
```
2)
```
from typing_extensions import Literal
# does require runtime dependency on typing_extensions
NormalizedDynamicFields = Literal[...]
```

Anyway, the only effect of the PR as it stands would be slightly looser type checking for those who use Python 3.7, so this is mostly a nit :-)

Reply:
Oh! I completely misinterpreted your comment and then when reading your example my confirmation bias took over and I was able to swear you were still using `from typing import Literal`... Sorry for that 😅

That sounds a very good approach to me, I will go ahead and implement it.

If the repository owners prefer keep the previous approach, please let me know and I will "revert" the commit.